File: BBDown.WebUI/src/lib/urlDetector.ts
/** 下载目标识别，复刻 GUI UrlDetector 的逻辑与文案（前缀来源与 Core IdPrefix 一致）。 */

/** 已知前缀，前缀后必须紧跟数字（BV 号固定以 BV1 开头，单独判定）。 */
const KNOWN_PREFIXES: [string, string][] = [
  ['av', '视频（av 号）'],
  ['ep', '番剧（ep 号）'],
  ['ss', '番剧（ss 号）'],
  ['md', '番剧（md 号）'],
  ['cheese/ep', '课程（ep 号）'],
  ['cheese/ss', '课程（ss 号）'],
  ['opus', '专栏（opus 号）'],
  ['cv', '专栏（cv 号）'],
  ['space', '用户空间'],
  ['live', '直播间（live 号）']
]

/** 识别输入文本，返回可读描述；无法识别返回 null。 */
export function describeTarget(input?: string): string | null {
  const text = input?.trim() ?? ''
  if (text.length === 0) {
    return null
  }

  const description = matchKnownPrefix(text)
  if (description !== null) {
    return description
  }

  if (/^[0-9]+$/.test(text)) {
    return '视频（av 号）'
  }

  try {
    const uri = new URL(text)
    if (uri.protocol === 'http:' || uri.protocol === 'https:') {
      return describeUrl(text)
    }
  } catch {
    // 非合法 URL，继续走 null
  }

  return null
}

function matchKnownPrefix(text: string): string | null {
  if (text.toLowerCase().startsWith('https://www.bilibili.com/watchlater')) {
    return '稍后再看列表'
  }

  // 裸 watchlater 简写与 URL 形态同义（与 Core InputResolver 的相等判定对齐）
  if (text.toLowerCase() === 'watchlater') {
    return '稍后再看列表'
  }

  if (text.toLowerCase().startsWith('https://live.bilibili.com')) {
    return '直播地址'
  }

  // BV 号固定以 BV1 开头且为纯 base58 字符，与 Core 的 bv1 前缀判定一致（BV2 等非 BV 号不应放行）
  if (/^bv1[0-9a-z]+$/i.test(text)) {
    return '视频（BV 号）'
  }

  for (const [prefix, label] of KNOWN_PREFIXES) {
    if (startsWithId(text, prefix)) {
      return label
    }
  }

  return null
}

function describeUrl(text: string): string {
  if (/\/cheese\//.test(text)) {
    return '课程地址'
  }

  const bv = /BV[0-9A-Za-z]+/.exec(text)
  if (bv?.[0]) {
    return `视频（${bv[0]}）`
  }

  // av / ep / ss 的路径形态为 .../video/av123、.../bangumi/play/ep123 等，关键字与数字直接相连
  if (/av[0-9]+/i.test(text)) {
    return '视频（av 号）'
  }

  if (/ep[0-9]+/i.test(text)) {
    return '番剧（ep 号）'
  }

  if (/ss[0-9]+/i.test(text)) {
    return '番剧（ss 号）'
  }

  // opus / cv 的路径形态为 .../opus/123...、.../cv/123...，关键字与数字间带斜杠（裸形态 opus123 同样成立）
  if (/opus\/?[0-9]+/i.test(text)) {
    return '专栏（opus 号）'
  }

  if (/cv\/?[0-9]+/i.test(text)) {
    return '专栏（cv 号）'
  }

  return '视频地址'
}

function startsWithId(text: string, prefix: string): boolean {
  if (text.length <= prefix.length || !text.toLowerCase().startsWith(prefix.toLowerCase())) {
    return false
  }

  return /\d/.test(text[prefix.length] ?? '')
}

File: BBDown.GUI/UrlDetector.cs
using System;
using System.Text.RegularExpressions;

using BBDown.Core;

namespace BBDown.GUI;

/// <summary>下载目标识别，纯函数；只描述识别结果，不做格式转换。ID 前缀复用 Core 的 IdPrefix 常量。</summary>
public static partial class UrlDetector
{
    /// <summary>识别输入文本，返回可读描述；无法识别返回 null。</summary>
    public static string? Describe(string? input)
    {
        var text = input?.Trim( ) ?? "";
        if (text.Length == 0)
        {
            return null;
        }

        if (MatchKnownPrefix(text) is { } description)
        {
            return description;
        }

        if (AvNumberRegex( ).IsMatch(text))
        {
            return "视频（av 号）";
        }

        if (Uri.TryCreate(text, UriKind.Absolute, out var uri) && uri.Scheme is "http" or "https")
        {
            return DescribeUrl(text);
        }

        return null;
    }

    /// <summary>匹配已知 ID 前缀与特殊 URL，前缀后必须紧跟数字（BV 号亦以数字开头）。</summary>
    private static string? MatchKnownPrefix(string text)
    {
        if (StartsWithId(text, IdPrefix.Av))
        {
            return "视频（av 号）";
        }

        if (StartsWithId(text, IdPrefix.Bv))
        {
            return "视频（BV 号）";
        }

        if (StartsWithId(text, IdPrefix.Ep))
        {
            return "番剧（ep 号）";
        }

        if (StartsWithId(text, IdPrefix.Ss))
        {
            return "番剧（ss 号）";
        }

        if (StartsWithId(text, IdPrefix.Md))
        {
            return "番剧（md 号）";
        }

        // 课程简写格式为 cheese/ep 号 / cheese/ss 号（与 Core 的 IdPrefix.CheeseSlash 前缀一致）
        if (StartsWithId(text, "cheese/ep"))
        {
            return "课程（ep 号）";
        }

        if (StartsWithId(text, "cheese/ss"))
        {
            return "课程（ss 号）";
        }

        if (StartsWithId(text, "opus"))
        {
            return "专栏（opus）";
        }

        if (StartsWithId(text, "cv"))
        {
            return "专栏（cv）";
        }

        if (StartsWithId(text, "space"))
        {
            return "用户空间";
        }

        // 集合简写（spaceOpus123 等）：space 分支要求 space 后紧跟数字，spaceOpus123 不会命中 space 分支
        if (StartsWithId(text, IdPrefix.SpaceOpus))
        {
            return "空间图文投稿";
        }

        if (StartsWithId(text, IdPrefix.SpaceAudio))
        {
            return "空间音频投稿";
        }

        if (StartsWithId(text, IdPrefix.SpaceDynamic))
        {
            return "空间动态";
        }

        if (StartsWithId(text, IdPrefix.ReadList))
        {
            return "文集";
        }

        if (StartsWithId(text, IdPrefix.Rl))
        {
            return "文集";
        }

        if (StartsWithId(text, IdPrefix.Au))
        {
            return "音频（au 号）";
        }

        if (StartsWithId(text, "live"))
        {
            return "直播间（live 号）";
        }

        if (text.StartsWith("https://www.bilibili.com/watchlater", StringComparison.OrdinalIgnoreCase))
        {
            return "稍后再看列表";
        }

        if (text.StartsWith("https://live.bilibili.com", StringComparison.OrdinalIgnoreCase))
        {
            return "直播地址";
        }

        return null;
    }

    private static string? DescribeUrl(string text)
    {
        if (text.Contains("/cheese/", StringComparison.OrdinalIgnoreCase))
        {
            return "课程地址";
        }

        if (text.Contains("/read/readlist/", StringComparison.OrdinalIgnoreCase))
        {
            return "文集地址";
        }

        // 空间子页限定 host（与 Core 的 TryParseCollection 守卫一致），非空间域的 /audio 等路径不误标
        var spaceHost = text.Contains("/space.bilibili.com/", StringComparison.OrdinalIgnoreCase);
        if (spaceHost && text.Contains("/upload/opus", StringComparison.OrdinalIgnoreCase))
        {
            return "空间图文投稿地址";
        }

        // 旧版音频页 space.bilibili.com/{mid}/audio 与新版 /upload/audio 同义（/audio 判定两者通吃）
        if (spaceHost && text.Contains("/audio", StringComparison.OrdinalIgnoreCase))
        {
            return "空间音频投稿地址";
        }

        if (spaceHost && text.Contains("/dynamic", StringComparison.OrdinalIgnoreCase))
        {
            return "空间动态地址";
        }

        // 单音频页 www.bilibili.com/audio/au12345（space 域的 /audio 列表页已在上面先行识别）
        if (text.Contains("/audio/au", StringComparison.OrdinalIgnoreCase))
        {
            return "音频地址（au 号）";
        }

        if (BvRegex( ).Match(text) is { Success: true } bv)
        {
            return $"视频（{bv.Value}）";
        }

        if (AvInUrlRegex( ).IsMatch(text))
        {
            return "视频（av 号）";
        }

        if (EpRegex( ).IsMatch(text))
        {
            return "番剧（ep 号）";
        }

        if (SsRegex( ).IsMatch(text))
        {
            return "番剧（ss 号）";
        }

        if (OpusRegex( ).IsMatch(text))
        {
            return "专栏（opus）";
        }

        if (CvRegex( ).IsMatch(text))
        {
            return "专栏（cv）";
        }

        return "视频地址";
    }

    private static bool StartsWithId(string text, string prefix)
    {
        if (text.Length <= prefix.Length || !text.StartsWith(prefix, StringComparison.OrdinalIgnoreCase))
        {
            return false;
        }

        return char.IsAsciiDigit(text[prefix.Length]);
    }

    [GeneratedRegex(@"^[0-9]+$")]
    private static partial Regex AvNumberRegex( );

    [GeneratedRegex(@"BV[0-9A-Za-z]+", RegexOptions.IgnoreCase)]
    private static partial Regex BvRegex( );

    [GeneratedRegex(@"av[0-9]+", RegexOptions.IgnoreCase)]
    private static partial Regex AvInUrlRegex( );

    [GeneratedRegex(@"ep[0-9]+", RegexOptions.IgnoreCase)]
    private static partial Regex EpRegex( );

    [GeneratedRegex(@"ss[0-9]+", RegexOptions.IgnoreCase)]
    private static partial Regex SsRegex( );

    [GeneratedRegex(@"opus/?[0-9]+", RegexOptions.IgnoreCase)]
    private static partial Regex OpusRegex( );

    [GeneratedRegex(@"cv/?[0-9]+", RegexOptions.IgnoreCase)]
    private static partial Regex CvRegex( );
}
